ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -144,225 +236,283 @@ chainSyncClient
+      :: IntersectionState blk tip
+      -> STM m (Maybe (IntersectionState blk tip))
+    intersectsWithCurrentChain is@IntersectionState { candidateFrag, currentFrag } = do
+      currentFrag' <- getCurrentChain
+      currentTip'  <- Our <$> getOurTip
+      if
+        | AF.headPoint currentFrag == AF.headPoint currentFrag ->
+          -- Our current chain didn't change, nothing to do
+          return $ Just is
+
+        | isJust (AF.intersectionPoint currentFrag' candidateFrag) ->
+          -- Our current chain changed, but it still intersects with candidate
+          -- fragment, so update the 'currentFrag' field and trim to the
+          -- candidate fragment to the same anchor point.
+          case AF.splitAfterPoint candidateFrag (AF.anchorPoint currentFrag') of
+           Nothing -> error
+             "anchor point must be on candidate fragment if they intersect"
+           Just (_, trimmedCandidateFrag) -> return $ Just is
+             { currentFrag   = currentFrag'
+             , candidateFrag = trimmedCandidateFrag
+             , currentTip    = currentTip'
+             }
+
+        | otherwise ->
+          -- No more intersection with the current chain
+          return Nothing
+
+    -- | Request the next message (roll forward or backward), unless our chain
+    -- has changed such that it no longer intersects with the candidate, in
+    -- which case we initiate the intersection finding part of the protocol.
+    --
+    -- Note that this is the only place we check whether our current chain has
+    -- changed.
+    nextStep :: IntersectionState blk tip
+             -> MkPipelineDecision
+             -> Nat n
+             -> Their tip
+             -> STM m (Consensus (ClientPipelinedStIdle n) blk tip m)
+    nextStep is mkPipelineDecision n theirTip =
+      intersectsWithCurrentChain is >>= \case
+        Just is' -> do
+          -- Our chain (tip) didn't change or if it did, it still intersects
+          -- with the candidate fragment, so we can continue requesting the
+          -- next block.
+          writeTVar varCandidate (candidateFrag is')
+          let candTipBlockNo = getTipBlockNo (unTheir theirTip)
+          return $ requestNext is' mkPipelineDecision n theirTip candTipBlockNo
+        Nothing ->
+          -- Our chain (tip) has changed and it no longer intersects with the
+          -- candidate fragment, so we have to find a new intersection, but
+          -- first drain the pipe.
+          drainThePipe n $ do
+            currentFrag       <- getCurrentChain
+            currentChainState <- ouroborosChainState <$> getCurrentLedger
+            ourTip            <- Our <$> getOurTip
+            let points = AF.selectPoints (map fromIntegral offsets) currentFrag
+                nis = NoIntersectionState
+                  { currentFrag       = currentFrag
+                  , currentChainState = currentChainState
+                  , currentTip        = ourTip
+                  }
```

Comment:
Duplication of initialization logic.

Reply:
(Above and below also, since we don't want to do the genesis special case anymore.)